Storing Points

craigtaverner opened the pull request:
This work provides initial support for storing Point geometries in the store. This involved some cleanup of the existing Point type support, and the PointValue system in particular. Some thinking has gone into future support for non-Point geometries, but only Point is supported at this point. Some key parts of this work:
- Use ValueWriter in PropertyStore for non-array types
- Use public interface for point in PointValue and contained types since the 3.3 removal of the runtime from the compiler model in Cypher
- Serialization/deserialization of 2D and 3D points (cannot support higher dimensions without moving to dynamic store, or switching precision to float, which is also considered in the format).

The key capabilities that would allow us to remove the 'not ready for merge' flag would be full stack support for assigning and reading Points as properties of nodes, using Cypher and Core API.

craigtaverner committed:
Removed unused ValueWriter method

craigtaverner committed:
Use ValueWriter for writing to PropertyStore

craigtaverner committed:
Reverted removal of one public API method, and deprecated it

sherfert committed:
Move PointValue to storable Value instead of VirtualValue

This commit prepares PointValues to be stored. the writePoint method
is not implemented in all places yet.

pontusmelke commented on community/kernel/src/main/java/org/neo4j/helpers/BaseToObjectValueWriter.java:
```diff
@@ -290,17 +285,43 @@ public void remove()
     }
 
-    @Override
+    @Deprecated
```
Just delete no need to deprecate

pontusmelke commented on community/values/src/main/java/org/neo4j/values/storable/PointValue.java:
```diff
@@ -0,0 +1,138 @@
+
+import static java.lang.String.format;
+
+public class PointValue extends ScalarValue
```
I know you badly want it but `Point` is not a storable value yet and shouldn't live here.

craigtaverner committed:
Removed deprecated methods after fix of API leak

craigtaverner committed:
Fixed checkstyle error

craigtaverner committed:
Fixed point type conversion for BaseToObjectValueWriter

sherfert committed:
Make PointValue comparable

tinwelint commented:
@craigtaverner there's no description on this PR. What is it about, what drove it, how does it affect things/storage? It's super hard to read a PR w/o getting primed with those sorts of contexts. Thanks.

sherfert committed:
Encode and decode points in th eproperty store.

craigtaverner committed:
Fixed geometry block count calculation to never throw exceptions

This is required for the page cache retry logic to work correctly.

craigtaverner committed:
Added working Cypher tests for writing and reading Points to properties

tinwelint commented:
How final is the format? Will it change soon or ongoing? This puts our versioning to the test... how do we version this?

craigtaverner commented:
@tinwelint The format is certainly not final. We are now at the stage where we would like feedback from kernel developers and anyone with interest in store format specification. The current state of the PR is a first draft of what we think makes sense, but should certainly be discussed further before finalizing.

sherfert committed:
Remove redundant PointValue factory methods.

craigtaverner committed:
Fixed failing test on print output

craigtaverner commented:
@tinwelint Discussions with @digitalstain last week indicate this format is backwards compatible (additive, upgrades require no store change) and partly forwards compatible (can downgrade as long as no-one added points to the store). We should decide how to (or if we need to) reflect this in the store version number.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/index/ArrayEncoder.java:
```diff
@@ -130,10 +131,17 @@ public void writeString( char value )
+        public void writePoint( CoordinateReferenceSystem crs, double[] coordinate ) throws RuntimeException
+        {
+            throw new UnsupportedOperationException( "Arrays of points are not yet supported." );
+            // Possible implementation:
```
this comment section should go if it's not supported

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java:
```diff
@@ -0,0 +1,164 @@
+                }
+            };
+
+    public final int gtype;
```
not sure why this field is public

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java:
```diff
@@ -0,0 +1,164 @@
+                    double[] coordinate = new double[dimension];
+                    for ( int i = 0; i < dimension; i++ )
+                    {
+                        coordinate[i] = Double.longBitsToDouble( valueBlocks[i + 1 + offset] );
```
no need to do +1 inside in the loop just do iteration from 1 till <= dimension

craigtaverner replied:
Then we need a `-1` inside the loop. Replacing `+1` with `-1` does not feel like a win to me.

MishaDemianenko replied:
you absolutely right, nevermind

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java:
```diff
@@ -0,0 +1,164 @@
+        this.gtype = gtype;
+    }
+
+    public abstract Value decode( CoordinateReferenceSystem crs, int dimension, long[] valueBlocks, int offset );
```
I can see that those 2 methods are abstract and we have only one type that overriding it. Do we plan to add more?

craigtaverner replied:
Yes :-)

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java:
```diff
@@ -0,0 +1,164 @@
+
+    private static int getGeometryType( long firstBlock )
+    {
+        return (int) ((firstBlock & 0x00000000F0000000L) >> 28);
```
all those magic numbers and masks should be extracted as constants with proper naming for future readers convenience

craigtaverner replied:
We copied the conventions we saw elsewhere in the code where it was easy to read the meaning of the value like this. If we change this place, perhaps we should change all other cases in PropertyStore where this happens?

MishaDemianenko replied:
if you have time to do that that would be really nice

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java:
```diff
@@ -0,0 +1,164 @@
+
+    public abstract Value decode( CoordinateReferenceSystem crs, int dimension, long[] valueBlocks, int offset );
+
+    // TODO this might not work this way for Polygons or other geometries that do not
```
who/when we should address this todo?

craigtaverner replied:
We would like to do this soon, but there are some unanswered questions around polygon indexing which we need to answer.

sherfert replied:
The answer is 1 for all Geometries storing in a dynamic store. That is also the default in PropertyType

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java:
```diff
@@ -0,0 +1,164 @@
+
+    private static GeometryType find( int gtype )
+    {
+        for ( GeometryType type : GeometryType.values() )
```
this for loop can be more effective in a couple of ways: "GeometryType.values()" are always the same so you do not need to create a new array each time; you can use your target gtype as an index to access required element directly without need to check each of arrays element for match in a for loop

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java:
```diff
@@ -0,0 +1,164 @@
+            throw new UnsupportedOperationException( "Points with more than 3 dimensions are not supported in the PropertyStore" );
+        }
+
+        long keyAndType = keyId | (((long) (PropertyType.GEOMETRY.intValue()) << StandardFormatSettings.PROPERTY_TOKEN_MAXIMUM_ID_BITS));
```
please use some local variable for StandardFormatSettings.PROPERTY_TOKEN_MAXIMUM_ID_BITS takes a lot of space and please described desired format somewhere in java doc

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java:
```diff
@@ -0,0 +1,164 @@
+        }
+        if ( dimension > 3 )
+        {
+            throw new UnsupportedOperationException( "Points with more than 3 dimensions are not supported in the PropertyStore" );
```
i would just say that "Points with more than 3 dimensions are not supported" since most of the supported/unsupported things are property format related and those can be quite different independently from type of store.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java:
```diff
@@ -0,0 +1,164 @@
+        int gtype = getGeometryType( firstBlock );
+        int dimension = getDimension( firstBlock );
+
+        if ( isFloatPrecision( firstBlock ) )
```
i can see this check is used only to say that it's unsupported. Is that how it was planned?

sherfert replied:
This is something open for discussion. We wanted to have the opportunity that users can trade precision for storage compactness. But we do not want to implement this right now.

MishaDemianenko replied:
if it will not be implemented right know what is the point of keeping that check?

sherfert replied:
As a reminder for later. Do you suggest to remove and simply assume it will be doublePrecision for now?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/PropertyStore.java:
```diff
@@ -226,91 +229,155 @@ public void encodeValue( PropertyBlock block, int keyId, Value value )
-        else if ( asObject instanceof Boolean )
+    }
+
+    private static class PropertyBlockValueWriter implements ValueWriter<IllegalArgumentException>
```
we tried to use property block value writer here before but benchmarks where showing degradation because of that please double check that this can be done now

MishaDemianenko replied:
This comment is still valid. Any benchmarks data?

sherfert replied:
Yes. No regressions from what we can see. An image is attached to the trello card.

MishaDemianenko commented on enterprise/runtime/neole/src/main/java/org/neo4j/internal/store/prototype/neole/PropertyCursor.java:
```diff
@@ -83,6 +84,12 @@
      *                                                                                 0 means 64, other values "normal"
      *            [xxxx,xxxx] [xxxx,xxxx] [xxxx,    ] [    ,    ] payload(+ maybe in next block) (0xFFFF_FF00_0000_0000)
      *                                                            bits are densely packed, bytes torn across blocks
+     * POINT:     [    ,    ] [    ,    ] [    ,    ] [    ,    ] [xxxx,type][K][K][K] geometry subtype
```
please move this format description to a place where that format is defined as well and reference this class or vica versa in javadoc

fickludd replied:
We are actually about to nuke NeoLE from the neo4j code base, so this is definitely not the future-proof place to document format changes. Maybe somewhere around the `PropertyRecordFormat`?

craigtaverner replied:
We should move all the format descriptions here to a better place. We only added the point format here because it was the only place where the formats are currently described in a reasonable way. Perhaps that move is done as part of the 'nuke neole' work?

sherfert replied:
Moved to PropertyStore

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/CRSTable.java:
```diff
@@ -17,36 +17,43 @@
+    CUSTOM( "custom", 0 ),
+    EPSG( "epsg", 1 ),
+    SR_ORG( "sr-org", 2 );
     public final String name;
```
same here as before why those field a republic?

craigtaverner replied:
I always use public fields for enums because they are immutable. Seems a common pattern, but we can do it differently with accessor methods if that is preferred for stylistic reasons.

MishaDemianenko replied:
yeah i would prefer access methods

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/CRSTable.java:
```diff
@@ -17,36 +17,43 @@
+    public static CRSTable find( int tableId )
     {
-        return name;
+        for ( CRSTable table : CRSTable.values() )
```
also same lookup pattern as described before

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/CoordinateReferenceSystem.java:
```diff
@@ -0,0 +1,102 @@
+    {
+        CRSTable table = CRSTable.find( tableId );
+        // TODO this might break, since we are changing the name. Use the function below?
+        return new CoordinateReferenceSystem( table.name + "-" + code, table, code );
```
not sure why do we need to create new instances of reference system here? we have only fixed number of those systems, right?

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/CoordinateReferenceSystem.java:
```diff
@@ -0,0 +1,102 @@
+
+import org.neo4j.graphdb.spatial.CRS;
+
+public class CoordinateReferenceSystem implements CRS
```
it really looks like this class should be enum instead?

craigtaverner replied:
There are a few thousand CRS out there, normally defined by libraries. We expect to either:
* Include an appropriate dependency to populate this table
* Allow users to include such dependencies to extend this table for their domain

Right now we will not expose that capability, and we could have an enum, but it feels better to have the code sympathetic to the direction we expect it to go.

MishaDemianenko replied:
ok sounds reasonable

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/Values.java:
```diff
@@ -253,6 +257,40 @@ public static ShortArray shortArray( short[] value )
+    {
+        // An optimization could be to do an instanceof PointValue check here
+        // and in that case just return the casted argument.
+        List<Double> coordinate = point.getCoordinate().getCoordinate();
```
".getCoordinate().getCoordinate()" looks quite strange

craigtaverner replied:
It does indeed. And sadly this is part of the public interface since 3.0. Personally I think that interface was not perfectly designed and would like to change it. In more ways than this one. Current policy is to only make changes on major releases. I wonder if we could, for a possibly entirely unused interface like this, motivate to make changes on minor releases? Worth discussing.

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/Values.java:
```diff
@@ -253,6 +257,40 @@ public static ShortArray shortArray( short[] value )
+
+    public static CoordinateReferenceSystem crs( CRS crs )
+    {
+        if ( crs.getHref().equals( CoordinateReferenceSystem.WGS84.href ) )
```
would prefer to have condition flipped to have constants on the left side to avoid NPEs in any case independently from crs that can be passed here ever

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/utils/PrettyPrinter.java:
```diff
@@ -152,16 +152,22 @@ public void writePath( NodeValue[] nodes, EdgeValue[] edges )
-        assert !stack.isEmpty();
-        append( stack.pop().done() );
+        append( "{geometry: {type: \"Point\", coordinates: [" );
+        for ( int i = 0; i < coordinate.length; i++ )
```
Arrays.toString( coordinate ) and you will have array [a, b, c] without need to put commas yourself and being careful about that

MishaDemianenko reviewed (changes requested):
In addition to already mentioned things would really like to see perf results for this change and also more integration tests with storing all sorts of coordinates: negative, big values, NANs if possible, etc

craigtaverner committed:
Fixed rebasing error

MishaDemianenko commented:
Also important note we need to bump store format versions in order to prevent people with older versions (without points) trying to read new property types

craigtaverner committed:
Corrected grammar on error message

craigtaverner committed:
Support PointArray and fixed spatial test

craigtaverner committed:
Support storing and retrieving arrays of points to/from properties

craigtaverner committed:
Added test case for new Bits.asBytes(offset) method

craigtaverner committed:
Considering some review comments

sherfert committed:
Adding GEOMETRY_INVALID type to avoid NullPointerExceptions.

sherfert committed:
Refactoring decodeArray into GeometryType

sherfert committed:
Considering some review comments.

sherfert committed:
Commenting the Property- and Dynamic Store formats.

Copied a comment from neole and also added description of the
DynamicArrayStore.

sherfert committed:
Fixed the bug reading point arrays from the store

The conversion from byte to int was wrong. Fixed it to use
Byte.toUnsignedInt instead.

sherfert committed:
Preventing storage of arrays of mixed CRS/mixed dimension points.

Added checks in both Kernel and Cypher, and some integration
tests for storing arrays of points.

sherfert committed:
Removed scala geometry type.

sherfert committed:
Fixing some small TODOs

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java:
```diff
@@ -0,0 +1,335 @@
+                }
+            };
+
+    public final int gtype;
```
comment about public field still valid

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java:
```diff
@@ -0,0 +1,335 @@
+            int geometryType = Byte.toUnsignedInt( header[1] );
+            int dimension = Byte.toUnsignedInt( header[2] );
+            int crsTableId = Byte.toUnsignedInt( header[3] );
+            int crsCode = Byte.toUnsignedInt( header[5] ) + (Byte.toUnsignedInt( header[4] ) << 8);
```
can we please update line to "(Byte.toUnsignedInt( header[4] ) << 8) + Byte.toUnsignedInt( header[5] ) "

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/TestArrayStore.java:
```diff
@@ -100,6 +103,15 @@ public void longArrayPropertiesShouldBeBitPacked() throws Exception
     @Test
+    public void doubleArrayPropertiesShouldBeBitPacked() throws Exception
+    {
+        assertBitPackedArrayGetsCorrectlySerializedAndDeserialized( new double[]{Double.longBitsToDouble( 0x1L ), Double.longBitsToDouble( 0x7L )},
```
please reformat those long lines in this test

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/CoordinateReferenceSystem.java:
```diff
@@ -0,0 +1,128 @@
+
+    public static CoordinateReferenceSystem get( CRS crs )
+    {
+        if ( crs == null || crs.getHref() == null )
```
since get by href is now null safe `crs.getHref()` is not mandatory anymore.

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/CoordinateReferenceSystem.java:
```diff
@@ -0,0 +1,128 @@
+    {
+        if ( crs == null || crs.getHref() == null )
+        {
+            throw new IllegalArgumentException( "Invalid CRS: " + crs );
```
maybe Objects.requireNonNull maybe use instead

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/CoordinateReferenceSystem.java:
```diff
@@ -0,0 +1,128 @@
+        }
+        else
+        {
+            throw new UnsupportedOperationException( "Unknown CRS: " + href );
```
would say that its more illegal argument

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/CoordinateReferenceSystem.java:
```diff
@@ -0,0 +1,128 @@
+        }
+    }
+
+    public final String name;
```
all of the fields should be private

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/CRSTable.java:
```diff
@@ -0,0 +1,97 @@
+        this.tableId = tableId;
+    }
+
+    private static final CRSTable[] TYPES = CRSTable.values();
```
please move all static stuff before a instance related stuff, here and in other new enums

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/storable/PointValue.java:
```diff
@@ -0,0 +1,203 @@
+    protected boolean eq( Object other )
+    {
+        return other != null &&
+               (
```
strange formatting

sherfert replied:
I actually cannot think of a more compact formatting that is still quite readable.

MishaDemianenko reviewed (changes requested):
Have couple of minor comments on my second round

sherfert committed:
Fixing review comments

fickludd commented on community/values/src/test/java/org/neo4j/values/storable/ValueComparisonTest.java:
```diff
@@ -38,6 +40,16 @@
+            pointValue( CoordinateReferenceSystem.Cartesian, -1.0, -1.0 ),
+            pointValue( CoordinateReferenceSystem.Cartesian, 1.0, 1.0 ),
+            pointValue( CoordinateReferenceSystem.Cartesian, 1.0, 2.0 ),
+            pointValue( CoordinateReferenceSystem.Cartesian, 2.0, 1.0 ),
```
Just got curious about this work... I think there should be point-arrays here somewhere too?

fickludd commented on community/values/src/test/java/org/neo4j/values/storable/PointTest.java:
```diff
@@ -0,0 +1,88 @@
+        assertEqual( pointValue( CoordinateReferenceSystem.Cartesian, 1.0, 2.0 ), pointValue( CoordinateReferenceSystem.Cartesian, 1.0, 2.0 ) );
+        assertEqual( pointValue( CoordinateReferenceSystem.Cartesian, -1.0, 2.0 ), pointValue( CoordinateReferenceSystem.Cartesian, -1.0, 2.0 ) );
+        assertEqual( pointValue( CoordinateReferenceSystem.Cartesian, -1.0, -2.0 ), pointValue( CoordinateReferenceSystem.Cartesian, -1.0, -2.0 ) );
+        assertEqual( pointValue( CoordinateReferenceSystem.Cartesian, 0.0, 0.0 ), pointValue( CoordinateReferenceSystem.Cartesian, 0.0, 0.0 ) );
```
This would be a nice spot for static import.

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/DynamicArrayStore.java:
```diff
@@ -34,18 +34,58 @@
+ *             <li>Points: Stored as double[] using the same format as primitive arrays above, starting with the 3 byte header (see above)</li>
+ *         </ul>
+ *     </li>
+ * </ul>
```
Nice!

sherfert committed:
Making PointArrays comparable and fixing minor review comments.

craigtaverner committed:
Created new 3_4 store format

sherfert committed:
Fix compilation errors from rebase

craigtaverner committed:
Follow pattern for number of longs in property block used by Points

sherfert committed:
Copy of High Limit Format

sherfert committed:
Test fixes

craigtaverner committed:
Fixed compile error after rebase

craigtaverner committed:
Spatial assignment does not work with 'morsel' runtime

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigratorTest.java:
```diff
@@ -205,6 +207,13 @@ public void shouldNotMigrateFilesForVersionsWithSameCapability() throws Exceptio
     }
 
+    @Test
+    public void shouldOpenStore()
```
what do we testing here?

sherfert committed:
Checking RecordFormats Version in PropertyStore and DynamicArrayStore

Broken tests. And, the check happens to late for the TX log.

MishaDemianenko committed:
Allow point and point arrays properties creation only in case underlying
record format support it, otherwise throw exception that will ask user to
upgrade store format to be able to use desired features.

MishaDemianenko committed:
Allow format selection based on available store without necessity
to perform upgrade to latest format always.

To allow users to use already available databases without necessity
of upgrade we will allow now automatic database format selection
during startup.
By introducing this feature we will allow users that can't perform
store format upgrade still to use new features and fixes that does not require
new store format.

MishaDemianenko committed:
Adapt tests

MishaDemianenko committed:
Return null in values module in case if can't recognise arrayValue.
Adapt record format related tests

MishaDemianenko committed:
Shutdown test databases

craigtaverner committed:
Merge remote-tracking branch 'upstream/3.4' into 3.4-property-store-values

craigtaverner committed:
Spatial now works with slotted

craigtaverner committed:
Fixed failing tests due to disabling double[] compaction
